src/github-stats.ts

```diff
@@ -0,0 +1,498 @@
+    Nemerle = '#3d3c6e',
+    nesC = '#94B0C7',
+    NetLinx = '#0aa0ff',
+    'NetLinx+ERB' = '#747faa',
+    NetLogo = '#ff6375',
+    NewLisp = '#87AED7',
+    Nextflow = '#3ac486',
+    Nim = '#37775b',
+    Nit = '#009917',
+    Nix = '#7e7eff',
+    Nu = '#c9df40',
+    'Objective-C' = '#438eff',
+    'Objective-C++' = '#6866fb',
+    'Objective-J' = '#ff0c5a',
+    OCaml = '#3be133',
+    Omgrofl = '#cabbff',
+    ooc = '#b0b77e',
+    Opal = '#f7ede0',
+    Oxygene = '#cdd0e3',
+    Oz = '#fab738',
+    P4 = '#7055b5',
+    Pan = '#cc0000',
+    Papyrus = '#6600cc',
+    Parrot = '#f3ca0a',
+    Pascal = '#E3F171',
+    PAWN = '#dbb284',
+    Pep8 = '#C76F5B',
+    Perl = '#0298c3',
+    'Perl 6' = '#0000fb',
+    PHP = '#4F5D95',
+    PigLatin = '#fcd7de',
+    Pike = '#005390',
+    PLSQL = '#dad8d8',
+    PogoScript = '#d80074',
+    PostScript = '#da291c',
+    PowerBuilder = '#8f0f8d',
+    PowerShell = '#012456',
+    Processing = '#0096D8',
+    Prolog = '#74283c',
+    'Propeller Spin' = '#7fa2a7',
+    Puppet = '#302B6D',
+    PureBasic = '#5a6986',
+    PureScript = '#1D222D',
+    Python = '#3572A5',
+    q = '#0040cd',
+    QML = '#44a51c',
+    R = '#198CE7',
+    Racket = '#22228f',
+    Ragel = '#9d5200',
+    RAML = '#77d9fb',
+    Rascal = '#fffaa0',
+    Rebol = '#358a5b',
+    Red = '#f50000',
+    'Ren\'Py' = '#ff7f7f',
+    Ring = '#0e60e3',
+    Roff = '#ecdebe',
+    Rouge = '#cc0088',
+    Ruby = '#701516',
+    RUNOFF = '#665a4e',
+    Rust = '#dea584',
+    SaltStack = '#646464',
+    SAS = '#B34936',
+    Scala = '#c22d40',
+    Scheme = '#1e4aec',
+    sed = '#64b970',
+    Self = '#0579aa',
+    Shell = '#89e051',
+    Shen = '#120F14',
+    Slash = '#007eff',
+    Smalltalk = '#596706',
+    Solidity = '#AA6746',
+    SourcePawn = '#5c7611',
+    SQF = '#3F3F3F',
+    Squirrel = '#800000',
+    'SRecode Template' = '#348a34',
+    Stan = '#b2011d',
+    'Standard ML' = '#dc566d',
+    SuperCollider = '#46390b',
+    Swift = '#ffac45',
+    SystemVerilog = '#DAE1C2',
+    Tcl = '#e4cc98',
+    Terra = '#00004c',
+    TeX = '#3D6117',
+    'TI Program' = '#A0AA87',
+    Turing = '#cf142b',
+    TypeScript = '#2b7489',
+    UnrealScript = '#a54c4d',
+    Vala = '#fbe5cd',
+    VCL = '#0298c3',
+    Verilog = '#b2b7f8',
+    VHDL = '#adb2cb',
+    'Vim script' = '#199f4b',
+    'Visual Basic' = '#945db7',
+    Volt = '#1F1F1F',
+    Vue = '#2c3e50',
+    wdl = '#42f1f4',
+    WebAssembly = '#04133b',
+    wisp = '#7582D1',
+    X10 = '#4B6BEF',
+    xBase = '#403a40',
+    XC = '#99DA07',
+    XQuery = '#5232e7',
+    XSLT = '#EB8CEB',
+    Yacc = '#4B6C4B',
+    Zephir = '#118f9e',
+}
+
+interface JsonCacheOptionsMap {
+    repos: Repo;
+    repoLanguages: RepoLanugages;
+}
+
+type JsonCacheType = keyof JsonCacheOptionsMap;
+
+type Language = keyof LANGUAGE_TO_COLOR;
+
+type LanguageCountMap = { [key in Language]?: number };
+
+export async function GithubStats(username: string): Promise<GithubStats> {
```

Comment:
Nit: I wonder whether we can make this an TypeScript class and abstract out all the I/O operations through interface and use constructor dependency injection to initialize this class. However, that may be complicated for the users to use.

Reply:
TODO: Ideally, in the future PRs, we want to separate caching, Github API, and SVG rendering into their own class to help simplify the logic and improves testability of the code:

```typescript
class GithubStats {
    private cache: Cache;
    private githubAPI: GithubAPI; // parsing the home page into JSON object and call Github languages APIs
    private diagramRender: GraphRender; // Render JSON object into diagrams on the Web
}
```

Let's together kill my legacy tech debts!

Reply:
I started with a similar idea, but it seems to be adding a lot of complexity to a simple thing. Maybe when we extend this further we can go ahead with that. For now just reorganizing and adding some typing.